codec-http/src/main/java/io/netty/handler/codec/http/HttpServerExpectContinueHandler.java

```diff
@@ -0,0 +1,83 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.util.ReferenceCountUtil;
+
+import static io.netty.handler.codec.http.HttpHeaderNames.CONTENT_LENGTH;
+import static io.netty.handler.codec.http.HttpResponseStatus.CONTINUE;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+
+/**
+ * The HttpServerExpectContinueHandler sends a <a href="https://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html#sec8.2.3">100
+ * CONTINUE</a> {@link HttpResponse} to {@link HttpRequest}s which contain a 'expect: 100-continue' header. It should
+ * only be used for applications which do <b>not</b> install the {@link HttpObjectAggregator}.
+ * <p>
+ * By default it accepts all expectations.
+ * <p>
+ * Since {@link HttpServerExpectContinueHandler} expects {@link HttpRequest}s it should be added after {@link
+ * HttpServerCodec} but before any other handlers that might send a {@link HttpResponse}. <blockquote>
+ * <pre>
+ *  {@link io.netty.channel.ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("respondExpectContinue", <b>new {@link HttpServerExpectContinueHandler}()</b>);
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+public class HttpServerExpectContinueHandler extends ChannelInboundHandlerAdapter {
+
+    private static final FullHttpResponse EXPECTATION_FAILED = new DefaultFullHttpResponse(
+            HTTP_1_1, HttpResponseStatus.EXPECTATION_FAILED, Unpooled.EMPTY_BUFFER);
+
+    static {
+        EXPECTATION_FAILED.headers().set(CONTENT_LENGTH, 0);
+    }
+
+    /**
+     * Returns {@code true} if the given {@link HttpRequest} should receive a 100 CONTINUE response.
+     * If {@code false} it will generate a 417 Expectation Failed {@link HttpResponse}.
+     */
+    protected boolean acceptMessage(HttpRequest request) {
+        return true;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (msg instanceof HttpRequest) {
+            HttpRequest req = (HttpRequest) msg;
+
+            if (HttpUtil.is100ContinueExpected(req)) {
+                if (acceptMessage(req)) {
+                    ctx.writeAndFlush(new DefaultFullHttpResponse(HTTP_1_1, CONTINUE));
+                    req.headers().remove(HttpHeaderNames.EXPECT);
+                    super.channelRead(ctx, msg);
+                } else {
+                    // the expection failed so we refuse the request with a final 417 response.
+                    ReferenceCountUtil.release(msg);
+                    ctx.writeAndFlush(EXPECTATION_FAILED.retainedDuplicate());
```

Comment:
@Scottmitch @jasontedor @normanmaurer overridable makes sense but 417 seems like it would be more a more informative default since it indicates the expectation as the error source, what's the reason to prefer 400 which describes "malformed syntax"?

The 417 as default behavior jives with the conclusion of the linked PR:

> That is, a server should respond with a 417 status if and only if there is an expectation that the server does not support (whether it be 100-continue, or another expectation extension), and should respond with another 4xx status code if the expectation is supported but there is something else wrong with the request.

Since if users override `acceptMessage` without the error code then by definition the server does not understand why it is rejecting a message other than it is expectation related.

Thoughts?